### PR TITLE
Bitcoin interpreter RPC service definition

### PR DESCRIPTION
**Pull request by jjayet-ledger**
## What is this about?

This is the RPC declaration of our Bitcoin Interpreter service.

## Cute picture of lama
great-pyrenees-with-lama

### CODE REVIEW

File: coins/bitcoin/interpreter/src/main/protobuf/service.proto [hunk 1]
```diff
@@ -3,12 +3,81 @@ syntax = "proto3";
...
+  int64 value = 2;
+  string address = 3;
+  string script_hex = 4;
+}
+
+message Block {
+  string hash = 1;
+  int32 height = 2;
+  string time = 3;
+}
```
Comment (vbergeron-ledger) [inline coins/bitcoin/interpreter/src/main/protobuf/service.proto:83]: new line

File: coins/bitcoin/interpreter/src/main/protobuf/service.proto [hunk 2]
```diff
@@ -3,12 +3,81 @@ syntax = "proto3";
...
 
+message Cursor {
+  string previous_page = 1;
+  string next_page = 2;
+}
+
+enum SortingOrder {
+  ASC = 0;
+  DESC = 1;
+}
```
Comment (vbergeron-ledger) [inline coins/bitcoin/interpreter/src/main/protobuf/service.proto:14]: `Cursor` object should have a `SortingOrder` ( that could be renamed as `CursorOrder` ) field.

- It allows the caller to just resend the returning cursor from the GET command
- A cursor without an order means nothing _per se_

File: coins/bitcoin/interpreter/src/main/protobuf/service.proto [hunk 3]
```diff
@@ -3,12 +3,81 @@ syntax = "proto3";
...
+
+message GetTransactionsResult {
+  repeated Transaction transactions = 1;
+  Cursor cursor = 2;
+}
+
+message Transaction {
+  string id = 1;
+  string hash = 2;
+  string received_at = 3;
```
Comment (vbergeron-ledger) [inline coins/bitcoin/interpreter/src/main/protobuf/service.proto:52]: Should be int64 (timestamp)

File: coins/bitcoin/interpreter/src/main/protobuf/service.proto [hunk 4]
```diff
@@ -3,12 +3,81 @@ syntax = "proto3";
...
+  int32 output_index = 1;
+  int64 value = 2;
+  string address = 3;
+  string script_hex = 4;
+}
+
+message Block {
+  string hash = 1;
+  int32 height = 2;
+  string time = 3;
```
Comment (vbergeron-ledger) [inline coins/bitcoin/interpreter/src/main/protobuf/service.proto:82]: should be int64 (timestamp)

**Review by vbergeron-ledger — CHANGES_REQUESTED** [review]

File: coins/bitcoin/interpreter/src/main/protobuf/service.proto (hunk 1 above)
Comment (jjayet-ledger) [reply coins/bitcoin/interpreter/src/main/protobuf/service.proto]: Fixed

File: coins/bitcoin/interpreter/src/main/protobuf/service.proto (hunk 2 above)
Comment (jjayet-ledger) [reply coins/bitcoin/interpreter/src/main/protobuf/service.proto]: This felt weird at first but I definitely see your point, thx for pointing that out!

File: coins/bitcoin/interpreter/src/main/protobuf/service.proto [hunk 5]
```diff
@@ -3,12 +3,81 @@ syntax = "proto3";
...
+  string script_signature = 6;
+  repeated string txinwitness = 7;
+  string sequence = 8;
+}
+
+message Output {
+  int32 output_index = 1;
+  int64 value = 2;
+  string address = 3;
+  string script_hex = 4;
```
Comment (onyb) [inline coins/bitcoin/interpreter/src/main/protobuf/service.proto:76]: I propose using the terminology [`ScriptPubKey`](https://learnmeabitcoin.com/technical/scriptPubKey) to follow the protocol's nomenclature.

```suggestion
  string script_pub_key = 4;
```

File: coins/bitcoin/interpreter/src/main/protobuf/service.proto [hunk 6]
```diff
@@ -3,12 +3,81 @@ syntax = "proto3";
...
+  SortingOrder sorting_order = 3;
+}
+
 service BitcoinInterpreterService {
-    rpc saveTransactions(Foo) returns (Foo) {}
-    rpc deleteTransactions(Foo) returns (Foo) {}
-    rpc getTransactions(Foo) returns (Foo) {}
+  rpc saveTransactions(SaveTransactionsRequest) returns (SaveTransactionsResult) {}
+  rpc deleteTransactions(DeleteTransactionsRequest) returns (DeleteTransactionsResult) {}
+  rpc getTransactions(GetTransactionsRequest) returns (GetTransactionsResult) {}
```
Comment (onyb) [inline coins/bitcoin/interpreter/src/main/protobuf/service.proto:20]: The convention is to keep the first letter uppercased, so `GetTransactions` as opposed to `getTransactions`.

**Review by onyb — COMMENTED** [review]
I have two main comments:

* Use canonical names for protobuf fields that have special meaning in the Bitcoin protocol. The names are somewhat oversimplified in the explorer, and we should use better names going forward.
* Consider using gRPC steams, so your RPC method would look like:
  ```
  rpc ListTransactions(GetTransactionsRequest) returns (stream Transaction) {}
  ```

**Review by vbergeron-ledger — APPROVED** [review]

**Pull request merged**